processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java

```diff
@@ -455,10 +456,17 @@ private DimensionSelector makeDimensionSelectorUndecorated(
                       }
 
                       if (columnDesc.getCapabilities().getType() == ValueType.LONG) {
+                        // Filtered dimension specs are not supported on numerics
+                        if (dimensionSpec instanceof BaseFilteredDimensionSpec) {
```

Comment:
Ugh, these `instanceof BaseFilteredDimensionSpec` checks everywhere are kind of gnarly. There's way too many of them for comfort. Is there something better we can do? Maybe making them work is a better option if it's not too hard… why don't they, again? They should be able to decorate a wrapped dimension selector just fine, right?

Is it just that we don't know if a DimensionSpec's "decorate" method is actually going to get called, because getColumnValueSelectorFromDimensionSpec might skip it by going and making something like a LongColumnSelector instead? If so, we could address that by adding a "mustDecorate" method to DimensionSpec, and if it mustDecorate, then we force creation of a DimensionSelector (rather than whatever type the underlying column is).

I would also accept alternate solutions that don't involve instanceof checks all over the place.